File: bash-server/clients/python/tests/test_integration.py
"""Integration tests requiring a real bash-server binary.

These tests start bash-server --stdio and run real commands.
Skip if bash-server is not available.
"""

import asyncio
import os
import shutil
import pytest

from bashclient.client import BashClient


BASH_SERVER = shutil.which("bash-server")
skip_no_server = pytest.mark.skipif(
    BASH_SERVER is None,
    reason="bash-server not found in PATH",
)

pytestmark = [pytest.mark.asyncio, skip_no_server]


class TestIntegration:
    async def test_stdio_connect_auth_eval(self):
        """Full lifecycle: connect, auth, eval, close."""
        proc = await asyncio.create_subprocess_exec(
            BASH_SERVER, "--stdio", "--token", "test" * 16,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        try:
            from bashclient.transport import StdioTransport
            transport = StdioTransport()
            transport._process = proc
            transport._reader = proc.stdout
            transport._writer_raw = proc.stdin
            transport._open = True

            client = await BashClient.from_transport(transport)
            await client.auth("test" * 16)
            result = await client.eval("echo integration_test")
            assert "integration_test" in result.stdout
            assert result.exit_code == 0
            await client.close()
        finally:
            try:
                proc.terminate()
                await asyncio.wait_for(proc.wait(), timeout=5.0)
            except Exception:
                proc.kill()

File: bash-server/clients/c/tests/test_integration.c
/*
 * test_integration.c - Integration test with real bash-server --stdio.
 * Skips if bash-server is not in PATH.
 */

#include "bashclient.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>

int main(void)
{
    alarm(30);

    /* Check if bash-server exists */
    if (system("which bash-server > /dev/null 2>&1") != 0) {
        printf("test_integration: SKIP (bash-server not found)\n");
        return 0;
    }

    const char *token = "tttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt";
    const char *argv[] = {"bash-server", "--stdio", "--token", token, NULL};

    bc_client_t *c = bc_connect_stdio(argv);
    if (!c) {
        printf("test_integration: FAIL (connect)\n");
        return 1;
    }

    int rc = bc_auth(c, token);
    if (rc != BC_OK) {
        printf("test_integration: FAIL (auth: %s)\n", bc_error(c));
        bc_close(c);
        return 1;
    }

    bc_eval_result_t result;
    rc = bc_eval(c, "echo integration_ok", &result);
    if (rc != BC_OK) {
        printf("test_integration: FAIL (eval: %s)\n", bc_error(c));
        bc_close(c);
        return 1;
    }

    if (!result.stdout_data || strstr(result.stdout_data, "integration_ok") == NULL) {
        printf("test_integration: FAIL (stdout: %s)\n",
               result.stdout_data ? result.stdout_data : "NULL");
        bc_eval_result_free(&result);
        bc_close(c);
        return 1;
    }

    printf("test_integration: PASS (stdout=%s exit=%d)\n",
           result.stdout_data, result.exit_code);
    bc_eval_result_free(&result);
    bc_close(c);
    return 0;
}
